### ISSUE TITLE
Request option for array-bracket-newline to only enforce adding line breaks, not removing of line breaks

### ISSUE THREAD

**Issue by om-mani-padme-hum** [eslint/eslint#12736]
**What rule do you want to change?**

array-bracket-newline

**Does this change cause the rule to produce more or fewer warnings?**

Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**

~Same defaults, but new default behavior~

New option for "multiline": true that only enforces adding of line breaks, not removing of line breaks.

**Please provide some example code that this change will affect:**

```js
indexes: [ /** Line 34 */
   { name: `serialNumber`, type: `BTREE`, columns: [ `serialNumber` ] }, 
 ]
```

**What does the rule currently do for this code?**

  34:12  error  There should be no linebreak after '['   array-bracket-newline
  36:3   error  There should be no linebreak before ']'  array-bracket-newline

**What will the rule do after it's changed?**

Not error just because there is a single line in array with line breaks after/before the [ and ].

**Are you willing to submit a pull request to implement this change?**

Sure

**Summary**

**Note:** The below comments are based on original issue, which has since been updated to reflect new option request instead of changes to default behavior.

A couple quick points:

1) The rule as-is assumes that one element is small, but multiple elements are large.  This is not necessarily true, one element may be an object which is large and looks way better on a line by itself. Meanwhile, 10 elements could be small enough to easily fit on one line (e.g. small integers), yet the rule has no problem with them being on separate lines.

2) If you read the configuration options for the rule, the "multiline": true option, with no other options set, SHOULD behave according to this proposed change based on the way it's described:

"multiline": true (default) requires line breaks if there are line breaks inside elements or between elements. If this is false, this condition is disabled.

**Note:** it says this condition is disabled if false, NOT that it errors when false.  See what I mean?  The rule seems meant to require new lines when there are line breaks inside elements or between elements, it's NOT saying new lines are not allowed for single elements.  Nor does it anywhere else, as the minItems rule is supposed to be ignored by default.

Thanks for your attention and for everything you guys do!

**Comment by kaicataldo** [comment]
Thanks for this proposal. The `multiline` option is a bit ambiguously named, but the intention seems to be that if the contents of the array are multiline, then it enforces newlines after and before the opening and closes brackets, respectively. Note that `multiline` doesn't check for the numbers of elements in the array - only that the elements are multiline. As a result, I don't think the default behavior of `multiline` should be changed.

I believe `minItems` is the option that is intended to address the style you want to enforce. Does setting `minItems: 1` fix this for you? If not, I would advocate for an additional option to be used in conjunction with `multiline` that makes the rule ignore arrays that contain a single element.

> it says this condition is disabled if false, NOT that it errors when false.

Regarding this wording, a PR to make the documentation clearer would be welcome!

**Comment by om-mani-padme-hum** [comment]
Thanks for the response!  Unfortunately, no, setting minItems: 1 makes any inline array flag as needed new lines.  e.g.

```js
const myArray = [1, 2, 3];
```

Now flags errors for needing new lines.

Either the multiline item is counting the number of elements in the array, and choosing to error when multiline arrays contain only one element, or the minItems option is misbehaving when it should be disabled and having no effect.

There is currently no solution that I've found using any combination of the options for this rule.

**Comment by mdjermanovic** [comment]
This is a 'multiline' array:

```js
[
  1,
  2
]
```

This isn't a 'multiline' array:

```js
[
  1, 2
]
```

I hope this helps to clarify the 'multiline' logic.

**Comment by mdjermanovic** [comment]
> Not error just because there is a single element in array

The rule actually isn't doing anything specific for single-element arrays. This would be also an error with the default settings (`"multiline: true"`):

```js
indexes: [ 
   { name: `serialNumber`, type: `BTREE`, columns: [ `serialNumber` ] }, { name: `serialNumber`, type: `BTREE`, columns: [ `serialNumber` ] } 
]
```

In my understanding, this is more a proposal to never disallow existing linebreaks after/before the brackets, in which case the `"consistent"` option might be the solution?

**Comment by om-mani-padme-hum** [comment]
It *is* doing something specific for single-element arrays.  It's throwing an error saying that I must change that to:

```js
indexes: [ { name: `serialNumber`, type: `BTREE`, columns: [ `serialNumber` ] }, { name: `serialNumber`, type: `BTREE`, columns: [ `serialNumber` ] } ]
```

Making the entire thing inline, but that's not a rule that I want, and is not the way the rule is described or seemingly intended.

The point of the rule as written seems to be to require line breaks after [ and before ] for multi-line arrays.  However, it is actually enforcing that you CAN'T have line breaks if there is only one element and that seems wrong, based on the way the rule is written.

```js
indexes: [
  { name: `serialNumber`, type: `BTREE`, columns: [ `serialNumber` ] }, { name: `serialNumber`, type: `BTREE`, columns: [ `serialNumber` ] }
 ]

// Errors
```

Unfortunately, `"consistent"` is too loose a rule for what I'm trying to accomplish.

**Comment by om-mani-padme-hum** [comment]
Per the docs:

`"multiline": true (default) requires line breaks if there are line breaks inside elements or between elements. If this is false, this condition is disabled.`

It is NOT meant to have the negative effect, which could be described as:

"multiline": true (default) requires that you DON'T use line breaks if there are not line breaks inside the elements or between elements in the array

**Comment by kaicataldo** [comment]
@om-mani-padme-hum While I agree that configuring style based on if the elements inside end up being on multiple lines isn't particularly intuitive, I do believe the rule is working as intended. Again, if the documentation isn't clear, it would be great to update it.

It sounds like you want `multiline` but to ignore arrays that only contain one element (allowing for both line breaks and no line breaks), correct? If so, my comment above still stands: 

> If not, I would advocate for an additional option to be used in conjunction with multiline that makes the rule ignore arrays that contain a single element.

**Comment by om-mani-padme-hum** [comment]
I simply want it to enforce line breaks when they are needed, and not error when line breaks are used when they are not needed.

**Comment by om-mani-padme-hum** [comment]
Perhaps you could say that "multiline" is incorrectly applying the "never" rule to single elements.

**Comment by kaicataldo** [comment]
Understood. I also understand why the word `"disabled"` is confusing in this context, and agree that the docs could be made clearer. The rule is written in a way that it always enforce newlines (or lack thereof), so this proposal will need to most likely include an option to allow that behavior.

**Comment by om-mani-padme-hum** [comment]
A new option plus improvement of documentation would work perfectly, yes, thanks!

**Comment by kaicataldo** [comment]
A PR to update documentation would be fantastic! For the rule changes, we'll want to hammer out a concrete proposal and will need to reach [consensus](https://eslint.org/docs/maintainer-guide/issues#consensus) within the team before we can move forward with it.

**Comment by om-mani-padme-hum** [comment]
Sounds good, thanks!  I'll work on the documentation PR this weekend.

**Comment by mdjermanovic** [comment]
> It _is_ doing something specific for single-element arrays. It's throwing an error saying that I must change that to:
> 
> ```js
> indexes: [ { name: `serialNumber`, type: `BTREE`, columns: [ `serialNumber` ] }  ]
> ```
> 

It would throw the same error for an array with two elements if they are on the same line.

[Online Demo Link](https://eslint.org/demo#eyJ0ZXh0IjoiaW5kZXhlczogWyBcbiAgIHsgbmFtZTogYHNlcmlhbE51bWJlcmAsIHR5cGU6IGBCVFJFRWAsIGNvbHVtbnM6IFsgYHNlcmlhbE51bWJlcmAgXSB9LCB7IG5hbWU6IGBzZXJpYWxOdW1iZXJgLCB0eXBlOiBgQlRSRUVgLCBjb2x1bW5zOiBbIGBzZXJpYWxOdW1iZXJgIF0gfSBcbl0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjgsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiYXJyYXktYnJhY2tldC1uZXdsaW5lIjoyfSwiZW52Ijp7fX19)

There is nothing specific in the behavior of this rule for single-element arrays. The rule works the same as it works with arrays that have more elements.

**Comment by om-mani-padme-hum** [comment]
Ah, I see what you mean, thanks for clarifying!  You're right, apparently I meant single-line arrays with new lines after/before the [ and ].  Still though, the desired functionality is not available given current options, nor does the documentation accurately describe the options as they are.  I hope we can find a way to allow the enforcement of new lines in multi-line arrays while not disallowing new lines in non-multi-line arrays.

**Comment by mdjermanovic** [comment]
As for the documentation, I think that it correctly describes the behavior:

> Or an object option (Requires line breaks if any of properties is satisfied. **Otherwise, disallows line breaks**)

Though, it does seem complicated to understand what's happening. If there is a better way to describe this, PR is welcome!

**Comment by om-mani-padme-hum** [comment]
That's true, thanks.  The documentation makes more sense now, but yes it's complicated, especially with the "disabled" comments on the "multiline" and "minItems" options.  Seems to be that either the caveat that "Otherwise, disallows line breaks" should be removed, or another option should be provided to only enforce adding of line breaks and not removing of line breaks.

Perhaps a documentation update is not needed, though could still be useful.  Nevertheless, I'd like to keep this issue open as a new option request if nothing else.  Thanks again for everyone's attention in this matter!

**Comment by mdjermanovic** [comment]
> Nevertheless, I'd like to keep this issue open as a new option request if nothing else. 

Of course, but please update the original post and the title with the new proposal when you have time. Or maybe it could be easier to open a new issue.

**Comment by om-mani-padme-hum** [comment]
Done, let me know if there's any other updates I should do.  I'll also still look at doing a PR for minor documentation improvements on the subject this weekend.

**Comment by om-mani-padme-hum** [comment]
Now that the documentation has been explained to me, it is actually pretty clear even though I misunderstood it.  Therefore, rather than do the documentation pull-request, I went ahead and forked the repository and implemented a new object option called "remove" for array-bracket-newline that is true by default (meaning original functionality) and if it's false, it no longer errors if line breaks are removed when not explicitly required.

Per your previous statements, this proposal requires consensus, and so I have not yet submitted a pull-request, but if you are interested in seeing my proposal in code, here is the forked repo with my changes to the rule, documentation, and tests.

https://github.com/om-mani-padme-hum/eslint

I'm fairly new to designing tests, so I'm not certain I've covered every circumstance required, but I tried to cover the same permutations as the original tests with the remove option set to false.  They are all currently passing.

I also signed the CLA in advance just in case the proposal is accepted.

**Comment by kaicataldo** [comment]
ESLint is also fully pluggable, so you can always also just create your own rules (without having to run your own forked version of ESLint). More info can be found [here](https://eslint.org/docs/developer-guide/working-with-plugins).

**Comment by om-mani-padme-hum** [comment]
> ESLint is also fully pluggable, so you can always also just create your own rules (without having to run your own forked version of ESLint). More info can be found [here](https://eslint.org/docs/developer-guide/working-with-plugins).

I'm not using it, I instead just put in /** eslint-disable-line */ comments in my files.  I just set it up so that it was ready for a pull-request if the consensus is reached to approve the addition of this option.  Thanks for the tip though, I will likely use that functionality in the future.
